src/stan-users-guide/parallelization.Rmd

```diff
@@ -261,35 +257,41 @@ model {
 
 ### Picking the Grainsize {#reduce-sum-grainsize}
 
-The `grainsize` is a recommendation on how large each piece of
+For `grainsize` is a recommendation on how large each piece of
 parallel work is (how many terms it contains). When using the
 non-static version, it is recommended to choose 1 as a starting
 point as automatic aggregation of partial sums are performed. However,
 for the static version the `grainsize` defines the maximal size of the
-partial sums, e.g. the static variant will split the input sequence
-until all partial sums are just smaller than `grainsize`. Therefore,
-for the static version it is more important to select a sensible
-value. The rational for choosing a sensible `grainsize` is based on
+partial sums, e.g. 
+
+The rational for choosing a sensible `grainsize` is based on
 balancing the overhead implied by creating many small tasks versus
 creating fewer large tasks which limits the potential parallelism.
 
-From empirical experience, the automatic grainsize determination works
-well and no further tuning is required in most cases. In order to
-figure out an optimal grainsize, think about how many terms are in the
-summation and on how many cores the model should run. If there are `N`
+In `reduce_sum`, `grainsize` is a recommendation on how to partition
+the work in the partial sum into smaller pieces. A `grainsize` of 1
+leaves this entirely up to the internal scheduler. Ideally this will be
+efficient, but there are no guarantees.
+
+In `reduce_sum_static`, `grainsize` is an upper limit on the worksize.
+Work will be split until all partial sums are just smaller than `grainsize`
+(and the split will happen the same way every time for the same data).
```

Comment:
I would rather speak of "the same data" of "the same input" to make clear that everything is meant (data and parameters).

Reply:
Ah, now I see your de-emphasis of my "let grainsize equal to 1" recommendation. I would actually really prefer to more strongly recommend people to just use grainsize = 1 and that's it. From what I have seen, this is well justified.

Reply:
Yeah that's how I'd like it to be, but I don't know if it is gonna turn out that way.

Honestly I'm not entirely sure about the behavior of grainsize with the auto partitioner. I'm happy to say, 'If nothing else, we recommend setting this to 1', but I don't see why we'd imply you shouldn't change this.

If you really shouldn't change this, it shouldn't be an option!

What are the rules on this with the auto partitioner. Is it something like the partitioner won't split stuff below this size? So it's sorta the opposite of the simple partitioner in that way?

Reply:
TBB doc for auto_partitioner:

> A loop template with an auto_partitioner attempts to minimize range splitting while providing ample opportunities for work-stealing.

> The range subdivision is initially limited to S subranges, where S is proportional to the number of threads specified by the task_scheduler_init or task_arena. Each of these subranges is not divided further unless it is stolen by an idle thread. If stolen, it is further subdivided to create additional subranges. Thus a loop template with an auto_partitioner creates additional subranges only when necessary to balance load.

> TIP
> When using auto_partitioner and a blocked_range for a parallel loop, the body may receive a subrange larger than the grain size of the blocked_range. Therefore do not assume that the grain size is an upper bound on the size of a subrange. Use simple_partitioner if an upper bound is required.

So the auto_partitioner splits according to how many threads you have available and it still provides possibilities for work stealing if that is possible. This will likely lead to larger blocks rather than smaller blocks for Stan programs.

Now, for Stan's autodiff the strategy to make it as large as possible is - that is what my hunch is - not always great. Once the work chunk size blows the CPU cache size, then this is really bad news for the reverse mode AD which is super memory intense.

Still, I would think that we confuse users a lot if we encourage to always tune it.

So we can settle on 'If nothing else, we recommend setting this to 1' - and this is where we one should start. People should concentrate on their problems, not on the grainsize.

Reply:
> So we can settle on 'If nothing else, we recommend setting this to 1'

Agreed. I'll make changes to that effect here in a few minutes.